### PR TITLE
More NIST curves

### PR DESCRIPTION
**Pull request by rubdos**
## Description

This PR implements secp256r1 and secp384r1; these are also known as NIST P-256 and NIST P-384 respectively. We are using these curves with Arkworks internally for research, and I thought it'd make sense to offer them as a PR, for others to use.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests (default set of unit tests)
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Review by Pratyush — COMMENTED** [review]
Thanks for the great PR! Please implement the remaining TODOs, and then this is good to go!

**Comment by rubdos** [comment]
Filled in the curve parameters in the docs in ac21ecf :)

File: secp384r1/src/lib.rs [hunk 1]
```diff
@@ -12,10 +12,12 @@
...
 //! Curve information:
-//! * Base field: q = TODO
-//! * Scalar field: r = TODO
-//! * a = TODO
-//! * b = TODO
+//! * Base field: q =
+//!   0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffeffffffff0000000000000000ffffffff
+//! * Scalar field: r =
+//!   0xffffffffffffffffffffffffffffffffffffffffffffffffc7634d81f4372ddf581a0db248b0a77aecec196accc52973
+//! * a = 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffeffffffff0000000000000000fffffffc
```
Comment (Pratyush) [inline secp384r1/src/lib.rs:19]: It seems like this is -3, right? Let's use that.
```suggestion
//! * a = -3
```

File: secp256r1/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+//! This library implements the secp256r1 curve.
+//! Source: <https://neuromancer.sk/std/secg/secp256r1>
+//!
+//! Curve information:
+//! * Base field: q =
+//!   0xffffffff00000001000000000000000000000000ffffffffffffffffffffffff
+//! * Scalar field: r =
+//!   0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
+//! * a = 0xffffffff00000001000000000000000000000000fffffffffffffffffffffffc
```
Comment (Pratyush) [inline secp256r1/src/lib.rs:19]:
```suggestion
//! * a = -3
```

File: secp256r1/src/curves/mod.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    /// COFACTOR_INV = COFACTOR^{-1} mod r = 1
+    #[rustfmt::skip]
+    const COFACTOR_INV: Fr =  Fr::ONE;
+}
+
+impl SWCurveConfig for Config {
+    /// COEFF_A = 115792089210356248762697446949407573530086143415290314195533631308867097853948
+    const COEFF_A: Fq =
+        MontFp!("115792089210356248762697446949407573530086143415290314195533631308867097853948");
```
Comment (Pratyush) [inline secp256r1/src/curves/mod.rs:33]:
```suggestion
    /// COEFF_A = -3
    const COEFF_A: Fq =
        MontFp!("-3");
```

File: secp384r1/src/curves/mod.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    /// COFACTOR_INV = COFACTOR^{-1} mod r = 1
+    #[rustfmt::skip]
+    const COFACTOR_INV: Fr =  Fr::ONE;
+}
+
+impl SWCurveConfig for Config {
+    /// COEFF_A = 39402006196394479212279040100143613805079739270465446667948293404245721771496870329047266088258938001861606973112316
+    const COEFF_A: Fq =
+        MontFp!("39402006196394479212279040100143613805079739270465446667948293404245721771496870329047266088258938001861606973112316");
```
Comment (Pratyush) [inline secp384r1/src/curves/mod.rs:33]:
```suggestion
    /// COEFF_A = -3
    const COEFF_A: Fq = MontFp!("-3");
```

**Review by Pratyush — APPROVED** [review]

File: secp384r1/src/lib.rs (hunk 1 above)
Comment (rubdos) [reply secp384r1/src/lib.rs]: That makes a lot of sense :)

**Pull request merged**